src/modules/musig/main_impl.h

```diff
@@ -0,0 +1,586 @@
+        secp256k1_ec_pubkey_serialize(ctx, commit, &commit_size, &session->nonce, SECP256K1_EC_COMPRESSED);
+        secp256k1_sha256_write(&sha, commit, commit_size);
+        secp256k1_sha256_finalize(&sha, nonce_commitment32);
+    }
+
+    secp256k1_scalar_clear(&secret);
+    return 1;
+}
+
+int secp256k1_musig_session_get_public_nonce(const secp256k1_context* ctx, secp256k1_musig_session *session, secp256k1_musig_session_signer_data *signers, secp256k1_pubkey *nonce, const unsigned char *const *commitments, size_t n_commitments) {
+    secp256k1_sha256 sha;
+    unsigned char nonce_commitments_hash[32];
+    size_t i;
+    (void) ctx;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(session != NULL);
+    ARG_CHECK(signers != NULL);
+    ARG_CHECK(nonce != NULL);
+    ARG_CHECK(commitments != NULL);
+
+    if (!session->has_secret_data || n_commitments != session->n_signers) {
+        return 0;
+    }
+    for (i = 0; i < n_commitments; i++) {
+        ARG_CHECK(commitments[i] != NULL);
+    }
+
+    secp256k1_sha256_initialize(&sha);
+    for (i = 0; i < n_commitments; i++) {
+        memcpy(signers[i].nonce_commitment, commitments[i], 32);
+        secp256k1_sha256_write(&sha, commitments[i], 32);
+    }
+    secp256k1_sha256_finalize(&sha, nonce_commitments_hash);
+    if (session->nonce_commitments_hash_is_set
+            && memcmp(session->nonce_commitments_hash, nonce_commitments_hash, 32) != 0) {
+        /* Abort if get_public_nonce has been called before with a different array of
+         * commitments. */
+        return 0;
+    }
+    memcpy(session->nonce_commitments_hash, nonce_commitments_hash, 32);
+    session->nonce_commitments_hash_is_set = 1;
+    memcpy(nonce, &session->nonce, sizeof(*nonce));
+    return 1;
+}
+
+int secp256k1_musig_session_initialize_verifier(const secp256k1_context* ctx, secp256k1_musig_session *session, secp256k1_musig_session_signer_data *signers, const unsigned char *msg32, const secp256k1_pubkey *combined_pk, const unsigned char *pk_hash32, const secp256k1_pubkey *pubkeys, const unsigned char *const *commitments, size_t n_signers) {
+    size_t i;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(session != NULL);
+    ARG_CHECK(signers != NULL);
+    ARG_CHECK(combined_pk != NULL);
+    ARG_CHECK(pk_hash32 != NULL);
+    ARG_CHECK(pubkeys != NULL);
+    ARG_CHECK(commitments != NULL);
+    for (i = 0; i < n_signers; i++) {
+        ARG_CHECK(commitments[i] != NULL);
+    }
+    (void) ctx;
+
+    memcpy(&session->combined_pk, combined_pk, sizeof(*combined_pk));
+    if (n_signers == 0) {
+        return 0;
+    }
+    session->n_signers = n_signers;
+    secp256k1_musig_signers_init(signers, pubkeys, n_signers);
+
+    memcpy(session->pk_hash, pk_hash32, 32);
+    session->nonce_is_set = 0;
+    session->msg_is_set = 0;
+    if (msg32 != NULL) {
+        memcpy(session->msg, msg32, 32);
+        session->msg_is_set = 1;
+    }
+    session->has_secret_data = 0;
```

Comment:
Maybe it's a good idea to memset the entire struct to zero in the beginning.

Reply:
you don't need it because you won't be able to call `get_public_nonce` with a verifier session, but I can be more explicit. Memsetting the struct seems like a good idea.

Reply:
Oh yes, you're right

Reply:
I'm adding the memset for both session types